### PR TITLE
fix: strict type check for cross realms

### PR DESCRIPTION
**Pull request by petamoriken**
Deno v1.39 introduces `vm.runInNewContext`. This may cause problems when using `Object.prototype.isPrototypeOf` to check built-in types.

```js
import vm from "node:vm";

const err = new Error();
const crossErr = vm.runInNewContext(`new Error()`);

console.assert( !(crossErr instanceof Error) );
console.assert( Object.getPrototypeOf(err) !== Object.getPrototypeOf(crossErr) );
```

This PR changes to check using internal slots solves them.

---

current: 

```
> import vm from "node:vm";
undefined
> vm.runInNewContext(`new Error("message")`)
Error {}
> vm.runInNewContext(`new Date("2018-12-10T02:26:59.002Z")`)
Date {}
```

this PR:

```
> import vm from "node:vm";
undefined
> vm.runInNewContext(`new Error("message")`)
Error: message
    at <anonymous>:1:1
> vm.runInNewContext(`new Date("2018-12-10T02:26:59.002Z")`)
2018-12-10T02:26:59.002Z
```

### CODE REVIEW

File: ext/web/08_text_encoding.js [hunk 1]
```diff
@@ -14,27 +14,52 @@ const ops = core.ops;
...
+
+function isSharedArrayBuffer(value) {
+  try {
+    getSharedArrayBufferByteLength(value);
+    return true;
+  } catch {
+    return false;
+  }
+}
+
```
Comment (littledivy) [inline ext/web/08_text_encoding.js:62]: Doing this across the codebase will be bad for perf and might also cause fast calls to deoptimize.

**Review by littledivy — COMMENTED** [review]
Thanks for looking into this. 

Taking into account the performance implications related to try/catch and .length calls, I suggest exploring an alternative approach. (Currently, vm.runInNewContext is solely utilized for Next.js support)

File: ext/web/08_text_encoding.js (hunk 1 above)
Comment (petamoriken) [reply ext/web/08_text_encoding.js:62]: Would it be faster to introduce `op_is_shared_array_buffer`?

Comment (littledivy) [reply ext/web/08_text_encoding.js:62]: Yes, that would be better 👍

**Comment by Martin978** [comment]
> # http[realista]
> rps	latencia	bytes	relativo
> nodo	86.376	114,31 µs σ=38,68	927.734 kB	-56,1%
> bollo	196.634	58,58 µs σ=121,52	1.835.937kB	mejor
> deno	143,514	68,53 µs σ=25,99	1.523.437kB	-27,0%
> línea de base deno	142.864	68,73 µs σ=25,62	1.513.671 kB	-27,3%
> # http[mínimo]
> rps	latencia	bytes	relativo
> nodo	86.670	113,65 µs σ=36,44	737.363 kB	-72,5%
> bollo	314.677	31,97 µs σ=16,02	1.552.734kB	mejor
> deno	171.009	56,87 µs σ=17,96	1.406.250 kB	-45,7%
> línea de base deno	171.540	57,11 µs σ=19,8	1.406.250 kB	-45,5%
> comenzar:`Fri, 22 Dec 2023 15:33:18 GMT`
> 
> identificación:`e648ce48-4fbd-4df4-8294-f1c54e177f73`
> 
> servidor:`72b7f325-a761-49fb-8931-d576283424d9`

**Comment by petamoriken** [comment]
We should move op_is_xxx to deno_core

File: Cargo.toml [hunk 2]
```diff
@@ -41,7 +41,7 @@ repository = "https://github.com/denoland/deno"
 
 [workspace.dependencies]
 deno_ast = { version = "0.31.6", features = ["transpiling"] }
-deno_core = { version = "0.239.0" }
+deno_core = { git = "https://github.com/denoland/deno_core", rev = "71a4589" }
```
Comment (petamoriken) [inline Cargo.toml:44]: Waiting for release

Comment (bartlomieju) [reply Cargo.toml:44]: I'll release it this week.

**Review by littledivy — APPROVED** [review]
LGTM, nice work. Hopefully no unknown perf regressions

**Comment by mmastrac** [comment]
Re-running benchmarks. I think we can merge this.

**Comment by littledivy** [comment]
/bench http[realistic,minimal]

The previous bench run probably crashed.

**Comment by littledivy** [comment]
Seems to be an issue with the bench bot. Merging.

**Pull request merged**